portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -26,21 +27,60 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-    private ArrayList<String> messages = new ArrayList<>(Arrays.asList("Message 1", "Message 2", "Message 3"));
-
-    /**
-     * Convert ArrayList as JSON using GSON library.
-    */
-    private String convertArrayListToJson(ArrayList<String> messages) {
-        Gson gson = new Gson();
-        String json = gson.toJson(messages);
-        return json;
-    }
+    private ArrayList<Comment> comments = new ArrayList<>();
 
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        response.setContentType("text/html;");
-        // response.getWriter().println("<h1>Hello Ariana!</h1>");
-        response.getWriter().println(convertArrayListToJson(messages));
+        response.setContentType("application/json");
+        String json = new Gson().toJson(comments);
+        response.getWriter().println(json);
+        System.out.println("doGet:");
+        System.out.println(json);
     }
+
+    @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        // Get input from the form
+        String name = getNameField(request);
+        String location = getLocationField(request);
+        String content = getContentField(request);
+        // TODO check which field this pops up
+        if (name == null || location == null || content == null) {
+            response.setContentType("text/html");
+            response.getWriter().println("Please fill in all fields.");
+            return;
+        }
+
+        comments.add(new Comment(name, location, content));
+
+        // Redirect back to HTML page
+        response.sendRedirect("/contact.html");
+  }
+
+  /** Returns the name field, or null if empty field. */
```

Comment:
I would suggest making this `getField(String fieldName)`, since the logic is the same for all of the `get*Field()` methods. So it is better not to repeat it in each function.

If you later decide to have special handling for particular fields, you can go back to something like `getNameField()`, but it would still want to do the common empty string to null handling via `getField("comment-name")`.